// @tableland/sdk/dist/cjs/helpers/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPromise = exports.camelize = exports.camelCase = exports.getDelay = exports.getRange = void 0;
function getRange(size, startAt = 0) {
    return [...Array(size).keys()].map((i) => i + startAt);
}
exports.getRange = getRange;
const getDelay = async (ms) => await new Promise((resolve) => setTimeout(resolve, ms));
exports.getDelay = getDelay;
function camelCase(str) {
    return str.replace(/[_.-](\w|$)/g, function (_, x) {
        return x.toUpperCase();
    });
}
exports.camelCase = camelCase;
function walk(obj) {
    if (obj == null || typeof obj !== "object")
        return obj;
    if (obj instanceof Date || obj instanceof RegExp)
        return obj;
    if (Array.isArray(obj))
        return obj.map(walk);
    return Object.keys(obj).reduce((res, key) => {
        const camel = camelCase(key);
        res[camel] = walk(obj[key]);
        return res;
    }, {});
}
function camelize(obj) {
    return typeof obj === "string" ? camelCase(obj) : walk(obj);
}
exports.camelize = camelize;
function isPromise(arg) {
    return Boolean(arg != null && typeof arg === "object" && typeof arg.then === "function");
}
exports.isPromise = isPromise;
